File: agent-go/edge_cases_test.go
package llmagent_test

import (
	"context"
	"encoding/json"
	"errors"
	"strings"
	"testing"

	llmagent "github.com/hoangvvo/llm-sdk/agent-go"
	llmsdk "github.com/hoangvvo/llm-sdk/sdk-go"
	"github.com/hoangvvo/llm-sdk/sdk-go/llmsdktest"
)

func TestRun_RejectsEmptyInputWithoutCallingModel(t *testing.T) {
	model := llmsdktest.NewMockLanguageModel()
	session := mustNewRunSession(
		t,
		&llmagent.AgentParams[struct{}]{Name: "test_agent", Model: model, MaxTurns: 10},
		struct{}{},
	)

	_, err := session.Run(t.Context(), llmagent.RunSessionRequest{})
	var agentErr *llmagent.AgentError
	if !errors.As(err, &agentErr) || agentErr.Kind != llmagent.InvariantErrorKind {
		t.Fatalf("expected invariant error, got %v", err)
	}
	if len(model.TrackedGenerateInputs()) != 0 {
		t.Fatal("model should not be called for empty input")
	}
}

func TestRun_RejectsDuplicateToolCallIDsBeforeExecution(t *testing.T) {
	model := llmsdktest.NewMockLanguageModel()
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{
			llmsdk.NewToolCallPart("duplicate", "first", map[string]any{}),
			llmsdk.NewToolCallPart("duplicate", "second", map[string]any{}),
		},
	}))
	first := NewMockTool[struct{}]("first", llmagent.AgentToolResult{}, nil)
	second := NewMockTool[struct{}]("second", llmagent.AgentToolResult{}, nil)
	session := mustNewRunSession(
		t,
		&llmagent.AgentParams[struct{}]{
			Name:     "test_agent",
			Model:    model,
			Tools:    llmagent.FunctionTools[struct{}](first, second),
			MaxTurns: 10,
		},
		struct{}{},
	)

	_, err := session.Run(t.Context(), llmagent.RunSessionRequest{Input: []llmagent.AgentItem{
		llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("Use tools"))),
	}})
	if err == nil || !strings.Contains(err.Error(), "duplicate tool call ID: duplicate") {
		t.Fatalf("expected duplicate-ID error, got %v", err)
	}
	if len(first.AllCalls) != 0 || len(second.AllCalls) != 0 {
		t.Fatal("tools must not execute after duplicate IDs are detected")
	}
}

func TestRun_PassesCurrentTurnAndAccumulatedItemsToTool(t *testing.T) {
	model := llmsdktest.NewMockLanguageModel()
	model.EnqueueGenerateResult(
		llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{Content: []llmsdk.Part{
			llmsdk.NewToolCallPart("call_1", "inspect_state", map[string]any{}),
		}}),
		llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{Content: []llmsdk.Part{
			llmsdk.NewTextPart("done"),
		}}),
	)
	type observation struct {
		turn      uint
		itemTypes []llmagent.AgentItemType
	}
	observed := make(chan observation, 1)
	tool := NewMockTool[struct{}]("inspect_state", llmagent.AgentToolResult{
		Content: []llmsdk.Part{llmsdk.NewTextPart("inspected")},
	}, func(_ context.Context, _ json.RawMessage, _ struct{}, state *llmagent.RunState) (llmagent.AgentToolResult, error) {
		items := state.Items()
		types := make([]llmagent.AgentItemType, 0, len(items))
		for _, item := range items {
			types = append(types, item.Type())
		}
		observed <- observation{turn: state.CurrentTurn, itemTypes: types}
		return llmagent.AgentToolResult{Content: []llmsdk.Part{llmsdk.NewTextPart("inspected")}}, nil
	})
	session := mustNewRunSession(
		t,
		&llmagent.AgentParams[struct{}]{
			Name: "test_agent", Model: model, Tools: llmagent.FunctionTools[struct{}](tool), MaxTurns: 10,
		},
		struct{}{},
	)

	_, err := session.Run(t.Context(), llmagent.RunSessionRequest{Input: []llmagent.AgentItem{
		llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("Inspect"))),
	}})
	if err != nil {
		t.Fatalf("run session: %v", err)
	}
	got := <-observed
	if got.turn != 1 || len(got.itemTypes) != 2 || got.itemTypes[0] != llmagent.AgentItemTypeMessage || got.itemTypes[1] != llmagent.AgentItemTypeModel {
		t.Fatalf("unexpected run-state observation: %#v", got)
	}
}

func TestRunStream_InvalidDeltaSequenceReturnsInvariantError(t *testing.T) {
	model := llmsdktest.NewMockLanguageModel()
	model.EnqueueStreamResult(llmsdktest.NewMockStreamResultPartials([]llmsdk.PartialModelResponse{
		{Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewTextPartDelta("hello")}},
		{Delta: &llmsdk.ContentDelta{Index: 0, Part: llmsdk.NewReasoningPartDelta("wrong type")}},
	}))
	session := mustNewRunSession(
		t,
		&llmagent.AgentParams[struct{}]{Name: "test_agent", Model: model, MaxTurns: 10},
		struct{}{},
	)
	stream, err := session.RunStream(t.Context(), llmagent.RunSessionRequest{Input: []llmagent.AgentItem{
		llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("Stream"))),
	}})
	if err != nil {
		t.Fatalf("create stream: %v", err)
	}
	for stream.Next() {
	}
	var agentErr *llmagent.AgentError
	if !errors.As(stream.Err(), &agentErr) || agentErr.Kind != llmagent.InvariantErrorKind {
		t.Fatalf("expected invariant stream error, got %v", stream.Err())
	}
}

func TestAgentRun_CancellationReachesToolAndStillClosesSession(t *testing.T) {
	model := llmsdktest.NewMockLanguageModel()
	model.EnqueueGenerateResult(llmsdktest.NewMockGenerateResultResponse(llmsdk.ModelResponse{
		Content: []llmsdk.Part{llmsdk.NewToolCallPart("call_1", "wait", map[string]any{})},
	}))
	started := make(chan struct{})
	tool := NewMockTool[struct{}]("wait", llmagent.AgentToolResult{}, func(ctx context.Context, _ json.RawMessage, _ struct{}, _ *llmagent.RunState) (llmagent.AgentToolResult, error) {
		close(started)
		<-ctx.Done()
		return llmagent.AgentToolResult{}, ctx.Err()
	})
	var cleanupContextErr error
	toolkitSession := &mockToolkitSession[struct{}]{
		closeFn: func(ctx context.Context) error {
			cleanupContextErr = ctx.Err()
			return nil
		},
	}
	agent := llmagent.NewAgent(
		"test_agent",
		model,
		llmagent.WithTools[struct{}](llmagent.NewAgentFunctionTool[struct{}](tool)),
		llmagent.WithToolkits[struct{}](&mockToolkit[struct{}]{createFn: func(context.Context, struct{}) (llmagent.ToolkitSession[struct{}], error) {
			return toolkitSession, nil
		}}),
	)
	ctx, cancel := context.WithCancel(t.Context())
	result := make(chan error, 1)
	go func() {
		_, err := agent.Run(ctx, llmagent.AgentRequest[struct{}]{
			Context: struct{}{},
			Input: []llmagent.AgentItem{
				llmagent.NewAgentItemMessage(llmsdk.NewUserMessage(llmsdk.NewTextPart("Wait"))),
			},
		})
		result <- err
	}()
	<-started
	cancel()
	err := <-result
	if !errors.Is(err, context.Canceled) {
		t.Fatalf("expected context cancellation, got %v", err)
	}
	if toolkitSession.closeCalls != 1 {
		t.Fatalf("expected toolkit cleanup after cancellation, got %d", toolkitSession.closeCalls)
	}
	if cleanupContextErr != nil {
		t.Fatalf("cleanup inherited canceled context: %v", cleanupContextErr)
	}
}

File: agent-js/src/edge-cases.test.ts
import { MockLanguageModel } from "@hoangvvo/llm-sdk/test";
import test, { suite, type TestContext } from "node:test";
import { Agent } from "./agent.ts";
import { AgentInvariantError } from "./errors.ts";
import { RunSession } from "./run.ts";
import type { AgentFunctionTool } from "./tool.ts";

suite("agent edge behavior", () => {
  test("rejects empty input without calling the model", async (t: TestContext) => {
    const model = new MockLanguageModel();
    const session = await RunSession.create({
      name: "test_agent",
      model,
      context: {},
    });

    await t.assert.rejects(
      () => session.run({ input: [] }),
      (error: unknown) => error instanceof AgentInvariantError,
    );
    t.assert.strictEqual(model.trackedGenerateInputs.length, 0);
  });

  test("rejects duplicate tool-call IDs before executing tools", async (t: TestContext) => {
    let executions = 0;
    const model = new MockLanguageModel();
    model.enqueueGenerateResult({
      response: {
        content: [
          {
            type: "tool-call",
            tool_call_id: "duplicate",
            tool_name: "first",
            args: {},
          },
          {
            type: "tool-call",
            tool_call_id: "duplicate",
            tool_name: "second",
            args: {},
          },
        ],
      },
    });
    const makeTool = (name: string): AgentFunctionTool<object> => ({
      type: "function",
      name,
      description: name,
      parameters: { type: "object", properties: {} },
      execute: () => {
        executions++;
        return { content: [], is_error: false };
      },
    });
    const session = await RunSession.create({
      name: "test_agent",
      model,
      tools: [makeTool("first"), makeTool("second")],
      context: {},
    });

    await t.assert.rejects(
      () =>
        session.run({
          input: [
            {
              type: "message",
              role: "user",
              content: [{ type: "text", text: "Use tools" }],
            },
          ],
        }),
      /Duplicate tool call ID: duplicate/,
    );
    t.assert.strictEqual(executions, 0);
  });

  test("exposes the current turn and accumulated items to tools", async (t: TestContext) => {
    const model = new MockLanguageModel();
    model.enqueueGenerateResult(
      {
        response: {
          content: [
            {
              type: "tool-call",
              tool_call_id: "call_1",
              tool_name: "inspect_state",
              args: {},
            },
          ],
        },
      },
      { response: { content: [{ type: "text", text: "done" }] } },
    );
    const inspectTool: AgentFunctionTool<object> = {
      type: "function",
      name: "inspect_state",
      description: "Inspect run state",
      parameters: { type: "object", properties: {} },
      execute: (_args, _context, state) => {
        t.assert.strictEqual(state.currentTurn, 1);
        t.assert.deepStrictEqual(
          state.getItems().map((item) => item.type),
          ["message", "model"],
        );
        return {
          content: [{ type: "text", text: "inspected" }],
          is_error: false,
        };
      },
    };
    const session = await RunSession.create({
      name: "test_agent",
      model,
      tools: [inspectTool],
      context: {},
    });

    const response = await session.run({
      input: [
        {
          type: "message",
          role: "user",
          content: [{ type: "text", text: "Inspect" }],
        },
      ],
    });
    t.assert.deepStrictEqual(response.content, [
      { type: "text", text: "done" },
    ]);
  });

  test("turns invalid streamed delta sequences into invariant errors", async (t: TestContext) => {
    const model = new MockLanguageModel();
    model.enqueueStreamResult({
      partials: [
        { delta: { index: 0, part: { type: "text", text: "hello" } } },
        {
          delta: {
            index: 0,
            part: { type: "reasoning", text: "wrong type" },
          },
        },
      ],
    });
    const session = await RunSession.create({
      name: "test_agent",
      model,
      context: {},
    });

    await t.assert.rejects(async () => {
      for await (const event of session.runStream({
        input: [
          {
            type: "message",
            role: "user",
            content: [{ type: "text", text: "Stream" }],
          },
        ],
      })) {
        t.assert.ok(event);
      }
    }, /Type mismatch at index 0/);
  });

  test("closing a one-shot stream early still cleans up its toolkit", async (t: TestContext) => {
    let closeCalls = 0;
    const model = new MockLanguageModel();
    model.enqueueStreamResult({
      partials: [
        { delta: { index: 0, part: { type: "text", text: "one" } } },
        { delta: { index: 0, part: { type: "text", text: "two" } } },
      ],
    });
    const agent = new Agent({
      name: "test_agent",
      model,
      toolkits: [
        {
          createSession: () =>
            Promise.resolve({
              getSystemPrompt: () => undefined,
              getTools: () => [],
              close: () => {
                closeCalls++;
                return Promise.resolve();
              },
            }),
        },
      ],
    });
    const stream = agent.runStream({
      context: {},
      input: [
        {
          type: "message",
          role: "user",
          content: [{ type: "text", text: "Stream" }],
        },
      ],
    });

    await stream.next();
    await stream.return({ content: [], output: [] });
    t.assert.strictEqual(closeCalls, 1);
  });
});
